server/mailet/integration-testing/src/test/java/org/apache/james/mailets/NetworkMatcherIntegrationTest.java

```diff
@@ -0,0 +1,320 @@
+                .matcherCondition("126.0.0.0/4")
+                .mailet(ToProcessor.class)
+                .addProperty("processor", "transport")
+                .build())
+            .addMailet(dropAllMailMailet()));
+
+        try (SMTPMessageSender messageSender =
+                 SMTPMessageSender.authentication(LOCALHOST_IP, SMTP_PORT, JAMES_APACHE_ORG, FROM, PASSWORD);
+             IMAPMessageReader imapMessageReader = new IMAPMessageReader(LOCALHOST_IP, IMAP_PORT)) {
+
+            messageSender.sendMessage(FROM, FROM);
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(messageSender::messageHasBeenSent);
+
+            calmlyAwait.atMost(Duration.ONE_MINUTE)
+                .until(() -> imapMessageReader.userReceivedMessage(FROM, PASSWORD));
+        }
+    }
+
+    @Test
+    public void mailsFromNonAuthorizedNetworksShouldNotBeDeliveredWithRemoteAddrInNetwork() throws Exception {
+        jamesServer = createJamesServerWithRootProcessor(ProcessorConfiguration.builder()
+            .state("root")
+            .addMailet(MailetConfiguration.builder()
+                .matcher(RemoteAddrInNetwork.class)
+                .matcherCondition("172.0.0.0/8")
+                .mailet(ToProcessor.class)
+                .addProperty("processor", "transport")
+                .build())
+            .addMailet(dropAllMailMailet()));
+
+        try (SMTPMessageSender messageSender =
+                 SMTPMessageSender.noAuthentication(LOCALHOST_IP, SMTP_PORT, JAMES_APACHE_ORG);
+             IMAPMessageReader imapMessageReader = new IMAPMessageReader(LOCALHOST_IP, IMAP_PORT)) {
+
+            messageSender.sendMessage(FROM, FROM);
+
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(messageSender::messageHasBeenSent);
+
+            Thread.sleep(TimeUnit.SECONDS.toMillis(10));
+
+            assertThat(imapMessageReader.userReceivedMessage(FROM, PASSWORD)).isFalse();
+        }
+    }
+
+    @Test
+    public void mailsFromNonAuthorizedNetworksShouldNotBeDeliveredWithRemoteAddrNotInNetwork() throws Exception {
+        jamesServer = createJamesServerWithRootProcessor(ProcessorConfiguration.builder()
+            .state("root")
+            .addMailet(MailetConfiguration.builder()
+                .matcher(RemoteAddrNotInNetwork.class)
+                .matcherCondition("127.0.0.0/8")
+                .mailet(ToProcessor.class)
+                .addProperty("processor", "transport")
+                .build())
+            .addMailet(dropAllMailMailet()));
+
+        try (SMTPMessageSender messageSender =
+                 SMTPMessageSender.noAuthentication(LOCALHOST_IP, SMTP_PORT, JAMES_APACHE_ORG);
+             IMAPMessageReader imapMessageReader = new IMAPMessageReader(LOCALHOST_IP, IMAP_PORT)) {
+
+            messageSender.sendMessage(FROM, FROM);
+
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(messageSender::messageHasBeenSent);
+
+            Thread.sleep(TimeUnit.SECONDS.toMillis(10));
```

Comment:
Why do we wait for a fixed time ?

Reply:
How do you ensure something did not happen?

I mean, you have to fix a duration after which you consider each environment would have performed the action, if it did it.

Yes, waiting for a side effect to happen is not clean, but :
 - i can not think about other strategies
 - we have always done like this.

However I have an alternative, but you won't like it, as it is less explicit in a test: define an alternative action, and wait for it.

Any opinion?

Reply:
Tommorrow I will instead wait on a **ToMailRepository** mailet condition.

This involves having a probe for accessing **MailRepository**.